programs/bpf_loader/src/serialization.rs

```diff
@@ -344,14 +352,23 @@ pub fn deserialize_parameters_aligned(
                 }
                 data_end
             };
-            let _ = borrowed_account.set_data(
-                buffer
-                    .get(start..data_end)
-                    .ok_or(InstructionError::InvalidArgument)?,
-            );
+            let data = buffer
+                .get(start..data_end)
+                .ok_or(InstructionError::InvalidArgument)?;
+            // The redundant `is_data_mutable` helps to avoid the expensive data comparison if we can
+            if borrowed_account.is_data_mutable(data.len()).is_ok()
+                || borrowed_account.get_data() != data
+            {
```

Comment:
As long as `enable_early_access_violation` is `false`, `is_data_mutable()` always returns `Ok(())` and this here evaluates to `if true`. So the behavior stays the same as it currently is.